File: ui/hybrid/sensitivity.js
/**
 * sensitivity.js — Yapı Tespit Hassasiyeti ve Kademe Dönüştürücü.
 *
 * Eşik katsayılarının TEK KAYNAĞI `shared/sensitivity.json` dosyasıdır;
 * Rust tarafı (`src-tauri/src/sensitivity.rs`) aynı dosyayı `include_str!`
 * ile okur. `golden` vektörleri iki taraftaki birim testlerle ortak doğrulanır.
 *
 * Kademe modeli:
 *   - ONAYLI (confirmed): yüksek oranlı tespitler (z-skor ≥ confirmedZ VEYA
 *     güven ≥ confirmedConf) hassasiyet eşiğinden MUAFTIR — çubuk kısalsa da
 *     asla silinmezler. Amaç: yüksek oranlı yapı ve anomalileri açığa çıkarmak.
 *   - ADAY (candidate): min_confidence / min_area aday eşiğini geçenler.
 *   - GÜRÜLTÜ (noise): kalanlar — listeden düşer, 3D'de soluk gösterilir.
 *
 * Çubuk ayrıca keşif tohumunu (seed_z) sürer: yüksek hassasiyet zayıf ama
 * tutarlı anomalileri de KEŞFEDER, düşük hassasiyet yalnız güçlü sinyali.
 */
import spec from "../../shared/sensitivity.json";

const MATCH_MIN = spec.match_threshold.min;
const MATCH_RANGE = spec.match_threshold.range;
const AREA_MAX = spec.min_area.max;
const AREA_RANGE = spec.min_area.range;
const CONF_MAX = spec.min_confidence.max;
const CONF_RANGE = spec.min_confidence.range;
const SEED_Z_MAX = spec.seed_z.max;
const SEED_Z_RANGE = spec.seed_z.range;
const CONFIRMED_Z = spec.confirmed.z;
const CONFIRMED_CONF = spec.confirmed.confidence;

/** Güven alanı 0–1'e normalize (0–100 ölçeğinde gelmiş olabilir). */
function confNorm(a) {
  if (typeof a?.confidence !== "number") return null;
  return a.confidence > 1 ? a.confidence / 100 : a.confidence;
}

/**
 * Hassasiyet yüzdesini (%0 - %100) teknik analiz eşiklerine dönüştürür.
 * @param {number} percent — 0 ile 100 arasında hassasiyet değeri
 * @returns {{
 *   percent: number,
 *   normalized: number,
 *   matchThreshold: number,
 *   minArea: number,
 *   minConfidence: number,
 *   seedZ: number,
 *   confirmedZ: number,
 *   confirmedConf: number,
 *   label: string,
 *   badgeColor: string,
 *   description: string
 * }}
 */
export function calculateSensitivityParameters(percent = 50) {
  const num = Number(percent);
  const p = Math.max(0, Math.min(100, Number.isFinite(num) ? num : 50));
  const norm = p / 100.0;

  // Renk/Sinyal Toleransı (0.15 katı .. 0.60 maksimum detay)
  const matchThreshold = Number((MATCH_MIN + norm * MATCH_RANGE).toFixed(3));

  // Min Yapı Piksel Alanı (250px kütlesel .. 15px ince detay)
  const minArea = Math.round(AREA_MAX - norm * AREA_RANGE);

  // ADAY güven eşiği (%80 katı .. %15 hassas sinyaller) — ONAYLI kademe muaf
  const minConfidence = Number((CONF_MAX - norm * CONF_RANGE).toFixed(2));

  // Keşif tohumu (σ): 2.5σ yalnız güçlü .. 1.0σ zayıf ama tutarlı anomaliler
  const seedZ = Number((SEED_Z_MAX - norm * SEED_Z_RANGE).toFixed(3));

  let label = "Dengeli";
  let badgeColor = "#eab308";
  let description = "Dengeli keşif · yüksek oranlılar ONAYLI, zayıflar ADAY";

  if (p <= 20) {
    label = "🔴 Katı (Düşük)";
    badgeColor = "#ef4444";
    description = "Katı aday eşiği · ONAYLI yüksek oranlı yapılar her koşulda görünür";
  } else if (p <= 45) {
    label = "🟠 Düşük";
    badgeColor = "#f97316";
    description = "Yalnız güçlü-orta yapılar aday olur · ONAYLI kademe muaf";
  } else if (p <= 65) {
    label = "🟡 Dengeli";
    badgeColor = "#eab308";
    description = "Dengeli keşif · yüksek oranlılar ONAYLI, zayıflar ADAY";
  } else if (p <= 85) {
    label = "🟢 Yüksek";
    badgeColor = "#3edc8c";
    description = "Zayıf anomali keşfi açık · tohum eşiği düşer, yeni yapılar açığa çıkar";
  } else {
    label = "🔵 Maksimum (Detaylı)";
    badgeColor = "#3b82f6";
    description = "Maksimum keşif · en zayıf tutarlı sinyaller bile aday kademeye girer";
  }

  return {
    percent: p,
    normalized: norm,
    matchThreshold,
    minArea,
    minConfidence,
    seedZ,
    confirmedZ: CONFIRMED_Z,
    confirmedConf: CONFIRMED_CONF,
    label,
    badgeColor,
    description,
  };
}

/**
 * Tespitin "oran" skoru (0–1): manyetik güç (z-skor) + güven bileşimi.
 * Yüksek oranlı tespitlerin sıralanması ve vurgulanması için kullanılır.
 * @param {Object} a — tespit nesnesi ({zScore, magnetic, confidence, area})
 * @returns {number} 0–1 arası oran skoru
 */
export function detectionScore(a = {}) {
  const strength =
    typeof a.zScore === "number"
      ? Math.min(1, Math.abs(a.zScore) / 4)
      : Math.min(1, Math.abs(a.magnetic || 0) / 500);
  const conf = confNorm(a) ?? 0.5;
  return Math.max(0, Math.min(1, 0.65 * strength + 0.35 * conf));
}

/**
 * Tek tespiti kademele: 'confirmed' | 'candidate' | 'noise'.
 * ONAYLI tespitler hassasiyet eşiğinden muaftır.
 * @param {Object} a — tespit nesnesi
 * @param {Object} params — calculateSensitivityParameters çıktısı
 * @returns {'confirmed'|'candidate'|'noise'}
 */
export function tierDetection(a, params) {
  const conf = confNorm(a);
  const zs = Math.abs(a.zScore || 0);
  if (zs >= params.confirmedZ || (conf != null && conf >= params.confirmedConf)) {
    return "confirmed";
  }
  const areaOk = typeof a.area !== "number" || a.area >= params.minArea;
  const judged = conf != null || typeof a.zScore === "number";
  if (!judged) return areaOk ? "candidate" : "noise";
  if (conf != null && conf >= params.minConfidence && areaOk) return "candidate";
  return "noise";
}

/**
 * Tespit listesini kademeleyip `tier` ve `score` alanlarıyla zenginleştirir.
 * @param {Array} anomalies
 * @param {number} sensitivityPercent — %0 - %100
 * @returns {Array} tier/score eklenmiş yeni liste (orijinal sıra korunur)
 */
export function annotateTiers(anomalies = [], sensitivityPercent = 50) {
  if (!Array.isArray(anomalies)) return [];
  const params = calculateSensitivityParameters(sensitivityPercent);
  return anomalies.map((a) => ({
    ...a,
    tier: tierDetection(a, params),
    score: detectionScore(a),
  }));
}

/**
 * Kademe kırılımı: {confirmed, candidate, noise, kept}.
 */
export function summarizeTiers(anomalies = [], sensitivityPercent = 50) {
  const tiers = annotateTiers(anomalies, sensitivityPercent);
  const counts = { confirmed: 0, candidate: 0, noise: 0, kept: 0 };
  for (const t of tiers) {
    counts[t.tier] += 1;
    if (t.tier !== "noise") counts.kept += 1;
  }
  return counts;
}

/**
 * Tespitleri hassasiyet seviyesine göre süzer — TEK KESİM NOKTASI.
 * ONAYLI (yüksek oranlı) tespitler her koşulda geçer; çubuk yalnız ADAY
 * eşiğini (min güven / min alan) uygular.
 * @param {Array} anomalies — Tespit edilen anomali/yapı nesneleri listesi
 * @param {number} sensitivityPercent — %0 - %100 hassasiyet
 * @returns {Array} ONAYLI + ADAY tespitler (kademe/score alanlarıyla)
 */
export function filterDetectionsBySensitivity(anomalies = [], sensitivityPercent = 50) {
  return annotateTiers(anomalies, sensitivityPercent).filter((t) => t.tier !== "noise");
}

/**
 * Min güven skorunu (0–1) tersine çevirip hassasiyet yüzdesine döndürür.
 * Arşiv geri yükleme gibi, kayıtlı min güven değerinden slider konumunu
 * yeniden kurmak için kullanılır. `calculateSensitivityParameters` içindeki
 * min güven formülünün tersidir ve 5'lik slider adımına yuvarlanır.
 * @param {number} minConfidence — 0.15 (hassas) ↔ 0.80 (katı)
 * @returns {number} 0–100 hassasiyet yüzdesi
 */
export function sensitivityPercentForMinConfidence(minConfidence = 0.45) {
  const raw = Number(minConfidence);
  const lo = CONF_MAX - CONF_RANGE;
  const mc = Math.max(lo, Math.min(CONF_MAX, Number.isFinite(raw) ? raw : 0.45));
  const pct = ((CONF_MAX - mc) / CONF_RANGE) * 100;
  return Math.max(0, Math.min(100, Math.round(pct / 5) * 5));
}

File: src-tauri/src/sensitivity.rs
//! Yapı tespit hassasiyeti (0.0 – 1.0) → teknik eşik eşlemesi.
//!
//! Katsayıların **tek kaynağı** `shared/sensitivity.json` dosyasıdır; JS tarafı
//! (`ui/hybrid/sensitivity.js`) aynı dosyayı import eder. `golden` vektörleri
//! iki taraftaki birim testlerle ortak doğrulanır — formül/katsayı sapması her
//! iki dilde de test hatasına düşer. Yuvarlama, JS'teki `toFixed` yuvarlamasıyla
//! aynı sonucu verir (0.5'ten yukarı).

use std::sync::OnceLock;

#[derive(Debug, serde::Deserialize)]
struct MatchThresholdCoeff {
    min: f32,
    range: f32,
}

#[derive(Debug, serde::Deserialize)]
struct CutCoeff {
    max: f32,
    range: f32,
}

#[derive(Debug, serde::Deserialize)]
struct ConfirmedSpec {
    z: f32,
    confidence: f32,
}

#[derive(Debug, serde::Deserialize)]
struct GoldenCase {
    percent: u32,
    match_threshold: f32,
    min_area: u32,
    min_confidence: f32,
    seed_z: f32,
}

#[derive(Debug, serde::Deserialize)]
struct Spec {
    match_threshold: MatchThresholdCoeff,
    min_area: CutCoeff,
    min_confidence: CutCoeff,
    seed_z: CutCoeff,
    confirmed: ConfirmedSpec,
    golden: Vec<GoldenCase>,
}

fn spec() -> &'static Spec {
    static SPEC: OnceLock<Spec> = OnceLock::new();
    SPEC.get_or_init(|| {
        serde_json::from_str(include_str!("../../shared/sensitivity.json"))
            .expect("shared/sensitivity.json çözümlenemeli")
    })
}

/// Renk/sinyal toleransı: 0.15 (katı) ↔ 0.60 (ince detay).
pub fn match_threshold(sensitivity: f32) -> f32 {
    let s = sensitivity.clamp(0.0, 1.0);
    let c = &spec().match_threshold;
    round3(c.min + s * c.range)
}

/// Min yapı piksel alanı: 250 px (kütlesel) ↔ 15 px (ince detay).
pub fn min_area(sensitivity: f32) -> u32 {
    let s = sensitivity.clamp(0.0, 1.0);
    let c = &spec().min_area;
    (c.max - s * c.range).round() as u32
}

/// Min güven skoru (ADAY eşiği): 0.80 (katı) ↔ 0.15 (hassas sinyaller).
/// ONAYLI kademe bu eşiğen muaftır (`is_confirmed`).
pub fn min_confidence(sensitivity: f32) -> f32 {
    let s = sensitivity.clamp(0.0, 1.0);
    let c = &spec().min_confidence;
    round2(c.max - s * c.range)
}

/// Keşif tohumu (σ): 2.5 (yalnız güçlü sinyal) ↔ 1.0 (zayıf ama tutarlı anomaliler).
/// Yüksek hassasiyet yeni anomalileri KEŞFEDER — sadece budamaz.
pub fn seed_z(sensitivity: f32) -> f32 {
    let s = sensitivity.clamp(0.0, 1.0);
    let c = &spec().seed_z;
    round3(c.max - s * c.range)
}

/// ONAYLI kademe z-skor eşiği (σ) — yüksek oranlı tespitler.
pub fn confirmed_z() -> f32 {
    spec().confirmed.z
}

/// ONAYLI kademe güven eşiği.
pub fn confirmed_confidence() -> f32 {
    spec().confirmed.confidence
}

/// ONAYLI (yüksek oranlı) yapı güveni: hassasiyet eşiğinden muaftır (hysteresis) —
/// çubuk kısalsa da silinmezler. Amaç: yüksek oranlı yapı ve anomalileri açığa çıkarmak.
pub fn is_confirmed(conf: f32) -> bool {
    conf >= confirmed_confidence()
}

fn round2(v: f32) -> f32 {
    (v * 100.0).round() / 100.0
}

fn round3(v: f32) -> f32 {
    (v * 1000.0).round() / 1000.0
}

#[cfg(test)]
mod tests {
    use super::*;

    fn approx(a: f32, b: f32) {
        assert!((a - b).abs() < 1e-6, "{a} ≉ {b}");
    }

    #[test]
    fn maps_zero_to_strict_thresholds() {
        approx(match_threshold(0.0), 0.15);
        assert_eq!(min_area(0.0), 250);
        approx(min_confidence(0.0), 0.80);
    }

    #[test]
    fn maps_midpoint_to_balanced_thresholds() {
        approx(match_threshold(0.5), 0.375);
        assert_eq!(min_area(0.5), 133);
        approx(min_confidence(0.5), 0.48);
    }

    #[test]
    fn maps_full_sensitivity_to_detail_thresholds() {
        approx(match_threshold(1.0), 0.60);
        assert_eq!(min_area(1.0), 15);
        approx(min_confidence(1.0), 0.15);
    }

    #[test]
    fn clamps_out_of_range_input() {
        approx(match_threshold(-1.0), 0.15);
        approx(match_threshold(2.0), 0.60);
        assert_eq!(min_area(-1.0), 250);
        assert_eq!(min_area(2.0), 15);
        approx(min_confidence(-1.0), 0.80);
        approx(min_confidence(2.0), 0.15);
    }

    #[test]
    fn seed_z_maps_discovery_thresholds() {
        approx(seed_z(0.0), 2.5);
        approx(seed_z(0.5), 1.75);
        approx(seed_z(1.0), 1.0);
        approx(seed_z(-1.0), 2.5);
        approx(seed_z(2.0), 1.0);
    }

    #[test]
    fn confirmed_tier_is_exempt_from_sensitivity_gate() {
        approx(confirmed_z(), 2.0);
        approx(confirmed_confidence(), 0.75);
        assert!(is_confirmed(0.92));
        assert!(is_confirmed(0.75));
        assert!(!is_confirmed(0.5));
        assert!(!is_confirmed(0.0));
    }

    #[test]
    fn golden_vectors_match_shared_spec() {
        for g in &spec().golden {
            let s = g.percent as f32 / 100.0;
            let mt = match_threshold(s);
            let ma = min_area(s);
            let mc = min_confidence(s);
            let sz = seed_z(s);
            assert!(
                (mt - g.match_threshold).abs() <= 0.001,
                "%{}: match_threshold {mt} ≠ {}",
                g.percent,
                g.match_threshold
            );
            assert_eq!(ma, g.min_area, "%{}: min_area", g.percent);
            assert!(
                (mc - g.min_confidence).abs() <= 0.01,
                "%{}: min_confidence {mc} ≠ {}",
                g.percent,
                g.min_confidence
            );
            assert!(
                (sz - g.seed_z).abs() <= 0.001,
                "%{}: seed_z {sz} ≠ {}",
                g.percent,
                g.seed_z
            );
        }
    }
}
